### PR TITLE
Added refresh rate for picture entity card

### PR DESCRIPTION
**Pull request by kkellner**
## Proposed change

The hui-image used in picture-entity card has a hard-coded refresh rate of 10 seconds.  To provide more flexibility in the image refresh rate a new user-configurable attribute has been added to the picture-entity card.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

For `picture-entity` card add a `refresh_rate` property.  This is optional and defaults to the original behavior of 10 seconds if the attribute is not defined.

```yaml
aspect_ratio: 70%
entity: camera.blueiris_cam24
show_name: false
show_state: false
type: picture-entity
refresh_rate: 2
```

## Additional information

- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/12710

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @kkellner, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;frontend#5466).

Once you do that we will be able to review and accept this pull request.

Thanks!

**Comment by homeassistant** [comment]
Hi @kkellner, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;frontend#5466).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: src/panels/lovelace/components/hui-image.ts [hunk 1]
```diff
@@ -173,7 +175,8 @@ export class HuiImage extends LitElement {
     if (this.cameraImage && this._attached) {
       this._cameraUpdater = window.setInterval(
         () => this._updateCameraImageSrc(),
-        UPDATE_INTERVAL
+        Number(this.refreshRate ? this.refreshRate : DEFAULT_REFRESH_RATE) *
+          1000
       );
```
Comment (bramkragten) [inline src/panels/lovelace/components/hui-image.ts:180]: Don't do the `Number` conversion, the values are numbers

File: src/panels/lovelace/components/hui-image.ts [hunk 2]
```diff
@@ -40,6 +40,8 @@ export class HuiImage extends LitElement {
 
   @property() public aspectRatio?: string;
 
+  @property() public refreshRate?: string;
```
Comment (bramkragten) [inline src/panels/lovelace/components/hui-image.ts:43]: This should be a number, not a string

File: src/panels/lovelace/components/hui-image.ts [hunk 3]
```diff
@@ -173,7 +175,8 @@ export class HuiImage extends LitElement {
     if (this.cameraImage && this._attached) {
       this._cameraUpdater = window.setInterval(
         () => this._updateCameraImageSrc(),
-        UPDATE_INTERVAL
+        Number(this.refreshRate ? this.refreshRate : DEFAULT_REFRESH_RATE) *
+          1000
```
Comment (bramkragten) [inline src/panels/lovelace/components/hui-image.ts:179]:
```suggestion
        (this.refreshRate || DEFAULT_REFRESH_RATE) * 1000
```

File: src/panels/lovelace/components/hui-image.ts [hunk 4]
```diff
@@ -208,9 +211,13 @@ export class HuiImage extends LitElement {
       return;
     }
 
+    const cacheTime =
+      (Number(this.refreshRate ? this.refreshRate : DEFAULT_REFRESH_RATE) - 1) *
```
Comment (bramkragten) [inline src/panels/lovelace/components/hui-image.ts:215]: Same as above

File: src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts [hunk 5]
```diff
@@ -36,6 +36,7 @@ const cardConfigStruct = struct({
   camera_image: "string?",
   camera_view: "string?",
   aspect_ratio: "string?",
+  refresh_rate: "string?",
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:39]: number

File: src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts [hunk 6]
```diff
@@ -79,6 +80,10 @@ export class HuiPictureEntityCardEditor extends LitElement
     return this._config!.aspect_ratio || "50";
   }
 
+  get _refresh_rate(): string {
+    return this._config!.refresh_rate || "10";
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:84]:
```suggestion
    return this._config!.refresh_rate || 10;
```

File: src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts [hunk 7]
```diff
@@ -79,6 +80,10 @@ export class HuiPictureEntityCardEditor extends LitElement
     return this._config!.aspect_ratio || "50";
   }
 
+  get _refresh_rate(): string {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:83]:
```suggestion
  get _refresh_rate(): number {
```

File: src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts [hunk 8]
```diff
@@ -188,6 +193,19 @@ export class HuiPictureEntityCardEditor extends LitElement
...
+          <paper-input
+            .label="${this.hass.localize(
+              "ui.panel.lovelace.editor.card.generic.refresh_rate"
+            )} (${this.hass.localize(
+              "ui.panel.lovelace.editor.card.config.optional"
+            )})"
+            type="number"
+            .value="${this._refresh_rate}"
+            .configValue="${"refresh_rate"}"
+            @value-changed="${this._valueChanged}"
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-picture-entity-card-editor.ts:206]:
```suggestion
            .value=${this._refresh_rate}
            .configValue=${"refresh_rate"}
            @value-changed=${this._valueChanged}
```

File: src/panels/lovelace/elements/hui-image-element.ts [hunk 9]
```diff
@@ -52,6 +52,7 @@ export class HuiImageElement extends LitElement implements LovelaceElement {
         .stateFilter="${this._config.state_filter}"
         .title="${computeTooltip(this.hass, this._config)}"
         .aspectRatio="${this._config.aspect_ratio}"
+        .refreshRate="${this._config.refresh_rate}"
```
Comment (bramkragten) [inline src/panels/lovelace/elements/hui-image-element.ts:55]:
```suggestion
        .refreshRate=${this._config.refresh_rate}
```

File: src/panels/lovelace/elements/types.ts [hunk 10]
```diff
@@ -46,6 +46,7 @@ export interface ImageElementConfig extends LovelaceElementConfigBase {
   filter?: string;
   state_filter?: string;
   aspect_ratio?: string;
+  refresh_rate?: string;
```
Comment (bramkragten) [inline src/panels/lovelace/elements/types.ts:49]:
```suggestion
  refresh_rate?: number;
```

File: src/translations/en.json [hunk 11]
```diff
@@ -2070,6 +2070,7 @@
             },
             "generic": {
               "aspect_ratio": "Aspect Ratio",
+              "refresh_rate": "Refresh Rate (seconds)",
```
Comment (bramkragten) [inline src/translations/en.json:2073]: Translations are handled in Localize, you should not add them here.

File: translations/ca.json [hunk 12]
```diff
@@ -2178,6 +2178,7 @@
                         },
                         "generic": {
                             "aspect_ratio": "Relació d'aspecte",
+                            "refresh_rate": "Velocitat d’actualització (segons)",
```
Comment (bramkragten) [inline translations/ca.json:2181]: Translations are handled in Localize, you should not add them here.

**Comment by bramkragten** [comment]
I don't think this is a good idea, a user can flood his Home Assistant now. We have `camera_view: live` for this.

File: src/panels/lovelace/components/hui-image.ts (hunk 1 above)
Comment (kkellner) [reply src/panels/lovelace/components/hui-image.ts:180]: I originally coded it with "number", however the value was never saved.  Maybe I missed somewhere I needed to change.

File: src/panels/lovelace/components/hui-image.ts (hunk 2 above)
Comment (kkellner) [reply src/panels/lovelace/components/hui-image.ts:43]: I originally coded it with "number", however the value was never saved. Maybe I missed somewhere I needed to change.

File: translations/ca.json (hunk 12 above)
Comment (kkellner) [reply translations/ca.json:2181]: This is my first time contributing to HA -- I didn't find any other finds that contained the translations.   When you say "handled in Localize", what do you mean?  Is there a different git repo for this?

Comment (zsarnett) [reply translations/ca.json:2181]: src\translations\en.json

This is the only place that it needs to be added

**Comment by kkellner** [comment]
The `live` view on my setup has a 7-12 second delay between showing on browser and what is "really live".   I am using Blue Iris -- when I go directly to the web interface of Blue Iris its immediate.  Not sure if there is buffering occuring in HA.   However, the image view (auto) was instant.   I agree that using low image refresh rates (e.g., 1 second) should be used with caution.

File: translations/ca.json (hunk 12 above)
Comment (kkellner) [reply translations/ca.json:2181]: Ok, I get it now.  Thanks for pointing that out.

**Comment by kkellner** [comment]
Changed everything to number and it is working.  The only thing that wasn't working well is when, for example, the user changed the default of 10 to 5, the user would press the delete key twice to delete "10" before entering "5".   When the field is empty it resulted in the field getting initialized back to 10.  I changed the getter to return NaN instead of 10 when value is not specified by user which then uses the internal default of 10.   Since this is an optional field, this seemed acceptable.

[NaN Change](https://github.com/home-assistant/frontend/pull/5466/commits/e46d0779a8cdaf8fcefbb79d6d4fba8597f496a4)

File: translations/ca.json (hunk 12 above)
Comment (zsarnett) [reply translations/ca.json:2182]: Please remove the other translation

Comment (kkellner) [reply translations/ca.json]: Removed

File: translations/ca.json [hunk 13]
```diff
@@ -2164,6 +2164,7 @@
                             "name": "Filtre d'entitats"
                         },
                         "entity": {
+                            "description": "La targeta entitat mostra una visualització ràpida dels estats d'una o més entitats.",
```
Comment (zsarnett) [inline translations/ca.json:2167]: All translations will need removed except for `src/translations/en.json`

Comment (kkellner) [reply translations/ca.json]: I believe I fixed.

Comment (zsarnett) [reply translations/ca.json]: Good now

**Comment by zsarnett** [comment]
Maybe we can limit this to being 10 seconds or greater? If the user makes it lower than 10 it is changed to being 10

**Comment by kkellner** [comment]
@zsarnett The original use case was to set the update interval to lower then 10 seconds as the live streaming from Blue Iris that flows through HA has a 7+ second delay.   So the work around was to refresh a static image every 3 seconds.   Maybe we can do some type of warning if the value is less than 10 seconds to inform user that the performance may be impacted.

**Comment by zsarnett** [comment]
Are you able to test and see if that work? Refreshing at a 3 second delay? IS HA now more in sync?

I know I see a delay with RTSP but I am not sure the delay is HA or just the RTSP

**Comment by kkellner** [comment]
I can have a 1 second refresh rate on a modern laptop and it keeps up, but the target is a Fire Tablet which is slow, so backed off to 3 seconds which seems to work well.

As for live streaming delay.   I can see that there is a 3 second delay coming from Blue Iris on the type of stream that is requested (m3u8 MIME type application/vnd.apple.mpegurl).  I think its because it has to transcode to get this format.   The additional delay seems to happen between between HA getting the stream and the browser getting the steaming bytes.  This is something I want to dig in more to see if the native "video/mpeg" content type can be used from Blue Iris as this is much faster -- but that's a different issue.

**Comment by bramkragten** [comment]
@kkellner I still don't think this is a good idea, if we would add this, it should not be in the UI editor as it is too advanced and could break systems.

Would it be an idea to create a custom card for this?

**Comment by kkellner** [comment]
@bramkragten It doesn't seem like a refresh rate is too advanced of an option.  Are you thinking that users will change to a value that is too low that HA would not be able to keep up?    I was trying to avoid creating another card type in order to keep the functionality together and the code in sync with what is already provided.  If you think this isn't going to workout, you can toss the change.

**Comment by nagyrobi** [comment]
3 seconds is perfect here. 10 seconds too slow.
I was looking for a way to speed up.

**Pull request closed without merging**